feat: add p2p URL support across all content in the reader

akhileshthite opened the pull request:
closes #17 

Facing two issues:

1. Getting null src for `img` and `video` in the post content. The src attributes might be empty or not set at the time when the script tries to update them?

```javascript
    const images = contentDOM.querySelectorAll('img')
    images.forEach(img => {
      const src = img.getAttribute('src')
      console.log('Original src:', src)
      img.setAttribute('src', resolveP2PUrl(src))
    })

    const videos = contentDOM.querySelectorAll('video source')
    videos.forEach(video => {
      const src = video.getAttribute('src')
      video.setAttribute('src', resolveP2PUrl(src))
    })
```

2. With the same updates in random sort `timelines` branch, getting this error:
![Screenshot 2024-05-26 at 3 13 29 AM](https://github.com/hyphacoop/reader.distributed.press/assets/68826419/2dc6a8fe-10ce-4a16-a273-c978ac6a9b57)

akhileshthite committed:
feat: add p2p URL support across all content in the reader

RangerMauve reviewed (changes requested):
We shouldn't be resolving to gateways in browsers that support p2p urls. Instead see if there's a DOM event that gets emitted when an image/video fails to load and listen to it at the `document.body` then take the `event.target` and attempt to rewrite its `src`. Video takes also have different ways of specifying sets of sources so we should account for that too.

RangerMauve commented:
For the `searchNotes` bit, check that you fixed the issues in #13 where the function got accidentally deleted

akhileshthite commented:
> We shouldn't be resolving to gateways in browsers that support p2p urls. Instead see if there's a DOM event that gets emitted when an image/video fails to load and listen to it at the `document.body` then take the `event.target` and attempt to rewrite its `src`. Video takes also have different ways of specifying sets of sources so we should account for that too.

Yep, will add the fallback logic.

This approach is not working:

```javascript
    // Setup error listeners for media elements within this component
    this.addEventListener('error', event => {
      const target = event.target;
      if (target.tagName === 'IMG' || (target.tagName === 'SOURCE' && target.parentElement.tagName === 'VIDEO')) {
        const originalSrc = target.getAttribute('src');
        if (!originalSrc.startsWith('hyper://') && !originalSrc.startsWith('ipns://')) {
          target.setAttribute('src', resolveP2PUrl(originalSrc));
        }
      }
    }, true);
  }
```

RangerMauve commented:
What part doesn't work? What steps have you taken to debug it so far?

akhileshthite commented:
> What part doesn't work? What steps have you taken to debug it so far?

I experimented with load events just to verify the functionality and discovered  that IPNS and HYPER URLs are resolving as `null`, including the `original/initial` URLs. It may be due to the specific timing of how we handle p2p URLs, particularly when elements are processed before they are fully loaded or accessible in the DOM?

Example code:-
```javascript
  connectedCallback() {
      // Listen for load events on the entire document body for images and video sources
      document.body.addEventListener('load', event => {
        const target = event.target;
    
        // Check if the loaded element is an image or a video source
        if (target.tagName === 'IMG' || (target.tagName === 'SOURCE' && target.closest('video'))) {
          const originalSrc = target.getAttribute('src');
    
          if (originalSrc && !originalSrc.startsWith('hyper://') && !originalSrc.startsWith('ipns://')) {
            const newSrc = resolveP2PUrl(originalSrc);
            console.log('Rewriting src for:', target, 'from:', originalSrc, 'to:', newSrc);
            target.setAttribute('src', newSrc);
          }
        }
      }, true);
  }
```

RangerMauve commented:
By resolving as null do you mean the `originalSrc` attribute is showing up as null or is the `resolveP2PURL` function returning null?

RangerMauve commented:
```
if (!originalSrc.startsWith('hyper://') && !originalSrc.startsWith('ipns://')) {
          target.setAttribute('src', resolveP2PUrl(originalSrc));
        }
```

This seems to only be trying to resolve the URL  if it *isn't* p2p. Should the if statement be resolved and have us try to update the `src` attribute right away?

Which browser are you testing this on? Which URL/post?

RangerMauve commented:
Mind committing your code?

akhileshthite commented:
> Mind committing your code?

Fixed it!
Will commit in a few hours.

akhileshthite committed:
Merge branch 'main' into resolve-p2p

akhileshthite committed:
fix: resolve p2p URLs before rendering to handle null src attributes

akhileshthite commented:
> > Mind committing your code?
> 
> Fixed it! Will commit in a few hours.

Resolving urls before rendering in `loadAndRenderPost` worked: https://github.com/hyphacoop/reader.distributed.press/pull/20/commits/675329fbd6051b2bb9d1b02200000842a771a962

Regarding the gateway fallback how is this approach in `db.js`?
```javascript
export async function supportsP2P (url) {
  console.log(url)
  try {
    const response = await fetch(url)
    return response.ok
  } catch (error) {
    console.log('P2P URL loading failed:', error)
    return false
  }
}

export async function resolveP2PUrl (url) {
  if (!url) return url

  console.log('Attempting to resolve P2P URL:', url)

  if (isP2P(url)) {
    const p2pSupported = await supportsP2P(url)

    if (!p2pSupported) {
      if (url.startsWith(HYPER_PREFIX)) {
        return url.replace(HYPER_PREFIX, 'https://hyper.hypha.coop/hyper/')
      } else if (url.startsWith(IPNS_PREFIX)) {
        return url.replace(IPNS_PREFIX, 'https://ipfs.hypha.coop/ipns/')
      }
    }
  }

  return url
}
```

RangerMauve commented:
Can you elaborate on what the purpose of the proposed changes to db.js are?

RangerMauve commented on actor-mini-profile.js:
```diff
@@ -50,7 +51,7 @@ class ActorMiniProfile extends HTMLElement {
     const img = document.createElement('img')
     img.className = 'profile-mini-icon'
-    img.src = iconUrl
+    img.src = resolveP2PUrl(iconUrl)
```
We shouldn't resolve to http unless there's an error loading. Maybe we should have a `p2p-image` custom element made which will attempt to render an `img` and fall back with the onerror handler? Then we can reuse the logic across pages.

RangerMauve commented on db.js:
```diff
@@ -103,6 +115,8 @@ export class ActivityPubDB extends EventTarget {
     if (url && typeof url === 'object') {
       return url
     }
+    url = resolveP2PUrl(url)
```
We shouldn't be trying to resolve to the gateway right away. Keep the current logic which resolve to the gateway only when the p2p logic fails.

RangerMauve commented on post.js:
```diff
@@ -39,6 +40,48 @@ function timeSince (dateString) {
+    console.log(`Set video src to: ${p2pSrc}`)
+  })
+
+  return contentDOM.body.innerHTML // Return the modified HTML to be inserted
```
Do the onerror events actually fire here? I think we need to add these listeners *after* we've sanitized and rendered the html for the post.

akhileshthite committed:
feat: implement media loading with p2p-image and p2p-video components

akhileshthite commented:
> Can you elaborate on what the purpose of the proposed changes to db.js are?

If the fetch fails, it will resolve the p2p URLs using gateways.

akhileshthite commented:
https://github.com/hyphacoop/reader.distributed.press/pull/20/commits/58366b88f0bbfa35b0515d5c904b71afc30e835f

The `p2p-media.js` is resolving to gateway if the fetch fails:
![Screenshot 2024-06-13 at 7 41 12 PM](https://github.com/hyphacoop/reader.distributed.press/assets/68826419/1820e6ce-0155-4e71-b21d-c2f85b4885eb)

The problem is it's not replacing the src:
original `img` src:
![Screenshot 2024-06-13 at 7 41 57 PM](https://github.com/hyphacoop/reader.distributed.press/assets/68826419/ae344cea-d277-4c24-8f01-c61a6d712756)

`p2p-image` src:
![Screenshot 2024-06-13 at 7 42 09 PM](https://github.com/hyphacoop/reader.distributed.press/assets/68826419/5dbbd0f3-9b0b-46de-9d96-17ee5ca7b9f7)

RangerMauve commented on p2p-media.js:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  handleError (source) {
+    const fallbackSrc = resolveP2PUrl(source.src)
```
Instead, check if the current `img.src` is p2p

RangerMauve commented on profile.html:
```diff
@@ -40,6 +40,7 @@
 <script type="module" src="./sidebar.js"></script>
 <script type="module" src="./actor-profile.js"></script>
 <script type="module" src="./post.js"></script>
+<script type="module" src="./p2p-media.js"></script>
```
We should have post.js import any of it's dependencies directly instead of doing it top level.

Similarly outbox should import post.

RangerMauve reviewed (changes requested):
We should modify the `supportsP2P` function so it only checks for the p2p schemes once.

something like this

- have a global map of `<protocol promise<boolean> | boolean>`
- supports will get the scheme from the src
- check if the map has a value for the scheme `if(supportMap.has(scheme)) return await supportMap.get(scheme)`
- if doesn't have, set a promise for attempting to resolve the url scheme
- if request fails on a network error (no http response) set it to false and return false
- else set to true

This will reduce the number of http requests and improve performance.

RangerMauve commented:
We already have a fallback for detecting when p2p fetch fails: https://github.com/hyphacoop/reader.distributed.press/blob/main/db.js#L116 Lets stick to that without adding new stuff

akhileshthite committed:
feat: implement dynamic CSS loading for p2p-media custom elements to ensure correct styling

feat: ensure class attributes are correctly synchronized for media elements

feat: optimize protocol support check with global caching

refactor: import module dependencies directly in respective scripts

akhileshthite commented:
`p2p-image` and `p2p-video` are working fine everywhere; the profile pictures in `actor-profile`, `actor-profile-mini`, and `followed-accounts`.

The media is not loading in the post content, I think there's something wrong with `insertImagesAndVideos`:
```
post.js:55 Updated p2p-image src after fallback: https://hyper.hypha.coop/hyper/hypha.coop/assets/images/posts/2022-05-12-ppl-sitting-unsplash.jpg
post.js:59 Replaced img with p2p-image having initial src: hyper://hypha.coop/assets/images/posts/2022-05-12-ppl-sitting-unsplash.jpg
```

I'm currently debugging the issue.

akhileshthite commented:
> We should modify the `supportsP2P` function so it only checks for the p2p schemes once.
> 
> something like this
> 
> * have a global map of `<protocol promise<boolean> | boolean>`
> * supports will get the scheme from the src
> * check if the map has a value for the scheme `if(supportMap.has(scheme)) return await supportMap.get(scheme)`
> * if doesn't have, set a promise for attempting to resolve the url scheme
> * if request fails on a network error (no http response) set it to false and return false
> * else set to true
> 
> This will reduce the number of http requests and improve performance.

https://github.com/hyphacoop/reader.distributed.press/blob/b5f63705addda30211318df4daf1988e66882aa7/db.js#L48

RangerMauve commented on p2p-media.js:
```diff
@@ -1,28 +1,43 @@
-    this.shadowRoot.appendChild(this.img)
+
+    // Load the CSS file
+    const style = document.createElement('style')
```
Couldn't we just add a regular style tag with `@mport` in the template?

TBH I think we should avoid the shadow root entirely and just embed the image into the plain dom and use regular css targetting the custom element.

RangerMauve replied:
Same for the video tag

RangerMauve commented on post.js:
```diff
@@ -40,42 +40,41 @@ function timeSince (dateString) {
     const p2pImg = document.createElement('p2p-image')
     p2pImg.setAttribute('src', originalSrc)
+
+    p2pImg.addEventListener('load-error', (event) => {
```
Do we need this here? Feels like we could encapsulate this logic entirely in the p2p image component.

RangerMauve commented on post.js:
```diff
@@ -40,42 +40,41 @@ function timeSince (dateString) {
     video.parentNode.replaceChild(p2pVideo, video)
   })
 
+  await Promise.all([...imgPromises, ...videoPromises])
```
These no longer need to be promises. We can just do for loops over the list of sources and append the new elements.

RangerMauve commented on post.js:
```diff
@@ -98,7 +97,7 @@ class DistributedPost extends HTMLElement {
       const content = await db.getNote(postUrl)
       if (content && content.content) {
-        content.content = insertImagesAndVideos(content.content) // Resolve URLs before rendering
+        content.content = await insertImagesAndVideos(content.content) // Resolve URLs before rendering
```
We should remove the async and just convert the img/video tags in a regular function since the load happens in the component.

sutty-coop commented:
@akhileshthite have you been able to apply th changes requested here?

akhileshthite commented:
> @akhileshthite have you been able to apply th changes requested here?

On it
Thanks

akhileshthite committed:
fix: ensure the p2p media elements are properly inserted and displayed within the post content

akhileshthite committed:
refactor: replace shadow DOM with plain DOM and update styling approach, chore: remove .vscode

akhileshthite committed:
fix: ensure default profile image is displayed when actor icon fails to load

akhileshthite commented:
Now we can load the p2p content in non p2p browsers :)

![Screenshot 2024-09-03 at 7 35 12 AM](https://github.com/user-attachments/assets/6645b58e-3d7c-496a-98bf-202ed5ce4ac6)
![Screenshot 2024-09-03 at 7 36 53 AM](https://github.com/user-attachments/assets/2d298d0f-689d-46f4-981a-8dde6b3ea060)

RangerMauve committed:
Merge branch 'main' into resolve-p2p

RangerMauve reviewed (approved):
Looks good, great job! Merge it in whenever
